File: JAVA/QuickSort.java
// This code is code for Quick Sort in Java.

import java.util.*;

class QuickSort {
    static int partition(List<Integer> arr, int low, int high) {
        int pivot = arr.get(low);
        int i = low;
        int j = high;

        while (i < j) {
            while (arr.get(i) <= pivot && i <= high - 1) {
                i++;
            }

            while (arr.get(j) > pivot && j >= low + 1) {
                j--;
            }
            if (i < j) {
                int temp = arr.get(i);
                arr.set(i, arr.get(j));
                arr.set(j, temp);
            }
        }
        int temp = arr.get(low);
        arr.set(low, arr.get(j));
        arr.set(j, temp);
        return j;
    }

    static void qs(List<Integer> arr, int low, int high) {
        if (low < high) {
            int pIndex = partition(arr, low, high);
            qs(arr, low, pIndex - 1);
            qs(arr, pIndex + 1, high);
        }
    }
    public static List<Integer> quickSort(List<Integer> arr) {
        // Write your code here.
        qs(arr, 0, arr.size() - 1);
        return arr;
    }
}

public class Sort {
    public static void main(String args[]) {
        List<Integer> arr = new ArrayList<>();
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        for (int i = 0; i < n; i++) {
            arr.add(scanner.nextInt());
        }
        arr = QuickSort.quickSort(arr);
        for (int i = 0; i < n; i++) {
            System.out.print(arr.get(i) + " ");
        }
        System.out.println();
    }

} 

File: sort_tech/quick_sort.c
#include <stdio.h>

// Function to partition the array and return the pivot index
int partition(int arr[], int low, int high, int* step) {
    int pivot = arr[high];  // Pivot element
    int i = low - 1;        // Index of the smaller element

    for (int j = low; j < high; j++) {
        // Print the array state before comparison
        printf("Step %d: ", (*step)++);
        for (int k = low; k <= high; k++) {
            printf("%d ", arr[k]);
        }

        // If the current element is smaller than or equal to the pivot
        if (arr[j] <= pivot) {
            i++;  // Increment index of the smaller element
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            printf("=> Swapped %d and %d", arr[i], arr[j]);
        }
        printf("\n");
    }

    // Swap the pivot element with the element at index (i + 1)
    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;

    printf("Step %d: ", (*step)++);
    for (int k = low; k <= high; k++) {
        printf("%d ", arr[k]);
    }
    printf("=> Swapped pivot %d and %d\n", arr[i + 1], arr[high]);

    return i + 1;  // Return the partition point
}

// Function to implement Quick Sort
void quickSort(int arr[], int low, int high, int* step) {
    if (low < high) {
        // Partition the array and get the pivot index
        int pivotIndex = partition(arr, low, high, step);

        // Recursively sort elements before and after partition
        quickSort(arr, low, pivotIndex - 1, step);
        quickSort(arr, pivotIndex + 1, high, step);
    }
}

int main() {
    int n;

    // Asking the user for the number of elements
    printf("Enter the number of elements: ");
    scanf("%d", &n);

    int arr[n];

    // Taking input from the user for the array
    printf("Enter the elements: ");
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    int step = 1;  // Step counter to track comparisons
    quickSort(arr, 0, n - 1, &step);

    // Final sorted array
    printf("\nSorted array: ");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    return 0;
}
